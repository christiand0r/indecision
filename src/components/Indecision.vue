<template>
  <img :src="image" alt="Background" />

  <div class="indecision-container">
    <input
      type="text"
      name="question"
      id="question"
      placeholder="Escribe tu pregunta..."
      v-model="question"
    />
    <p>Recuerda terminar con un signo de interrogación (?)</p>

    <div>
      <h2>{{ addSymbol }}</h2>
      <h1 v-if="isValidQuestion">{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  name: "Indecision",
  data() {
    return {
      question: "",
      answer: null,
      image: "https://media.giphy.com/media/aFTt8wvDtqKCQ/giphy.gif",
      isValidQuestion: false,
    };
  },
  watch: {
    question(value) {
      this.isValidQuestion = false;

      if (!value.includes("?")) return;

      this.isValidQuestion = true;

      this.getAnswer();
    },
  },
  methods: {
    async getAnswer() {
      try {
        //Hash Table
        const mapAnswer = {
          no: [
            "Definitivamente no",
            "No!",
            "Para nada",
            "Ni siquiera lo pienses",
            "Ni soñando",
            "No te hagas ilusiones",
            "Nunca!",
            "Jamás",
            "Imposible",
          ],
          yes: [
            "Claro que si",
            "Por supuesto",
            "Indudablemente",
            "Obvio bobis",
            "No lo dudes",
            "Shi señor!",
            "Si!",
            "Definitivamente",
            "Súper si",
          ],
          maybe: [
            "Con esfuerzo y dedicación",
            "El qué persevera alcanza",
            "Quizás",
            "No es claro... pero si seguro",
            "Posiblemente",
            "Habrá que averiguar",
            "No tengo pruebas, pero si dudas",
          ],
        };

        this.answer = "Pensado...";

        const { answer, image } = await fetch("https://yesno.wtf/api").then(
          (res) => res.json()
        );

        //Obtenemos un número random siendo el máximo el largo del array
        const position = this.getRandomNumber(mapAnswer[answer].length);

        this.answer = mapAnswer[answer][position];

        this.image = image;
      } catch (error) {
        console.log(error);
      }
    },
    getRandomNumber(max) {
      return Math.floor(Math.random() * max + 1);
    },
  },
  computed: {
    addSymbol() {
      if (this.question.includes("¿")) return this.question;

      let arrayWords = this.question.split("");
      arrayWords.unshift("¿");
      return arrayWords.length > 1 ? arrayWords.join("") : "";
    },
  },
};
</script>

<style scoped>
img {
  height: 100%;
  left: 0;
  object-fit: cover;
  opacity: 0.2;
  position: fixed;
  top: 0;
  width: 100%;
}

input {
  border-radius: 0.5rem;
  border: none;
  font-family: inherit;
  font-size: 1.2rem;
  max-width: 650px;
  padding: 1rem 1.5rem;
  width: 100%;
}

input:focus {
  outline: none;
}

h1,
h2,
p {
  color: white;
}

h2 {
  font-size: 3rem;
  margin-top: 10rem;
}

p {
  font-size: 20px;
  margin: 0.5rem 0 0 0;
}

.indecision-container {
  display: flex;
  align-items: center;
  flex-direction: column;
  height: 100vh;
  justify-content: center;
  position: relative;
  text-align: center;
  z-index: 2;
}
</style>
