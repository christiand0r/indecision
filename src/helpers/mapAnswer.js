export default {
  no: [
    "Definitivamente no",
    "No!",
    "Para nada",
    "Ni siquiera lo pienses",
    "Ni soñando",
    "No te hagas ilusiones",
    "Nunca!",
    "Jamás",
    "Imposible",
  ],
  yes: [
    "Claro que si",
    "Por supuesto",
    "Indudablemente",
    "Obvio bobis",
    "No lo dudes",
    "Shi señor!",
    "Si!",
    "Definitivamente",
    "Súper si",
  ],
  maybe: [
    "Con esfuerzo y dedicación",
    "El qué persevera alcanza",
    "Quizás",
    "No es claro... pero si seguro",
    "Posiblemente",
    "Habrá que averiguar",
    "No tengo pruebas, pero si dudas",
  ],
};
