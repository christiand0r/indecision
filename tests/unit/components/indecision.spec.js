import { shallowMount } from "@vue/test-utils";
import mapAnswer from "@/helpers/mapAnswer";
import Indecision from "@/components/Indecision.vue";

describe("Indecision Component", () => {
  let wrapper = null;

  let clogSpy = null;
  let getAnswerSpy = null;

  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: "yes",
          forced: false,
          image: "https://yesno.wtf/assets/yes/2.gif",
        }),
    })
  );

  beforeEach(() => {
    wrapper = shallowMount(Indecision);

    //Set Spy for method getAnswer
    getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");

    jest.clearAllMocks();
  });

  test("debe ser igual al snapshot", () =>
    expect(wrapper.html()).toMatchSnapshot());

  test("escribir en el input no debe ejecutar nada", async () => {
    const input = wrapper.find(".indecision-container input");

    await input.setValue("Ask a question");

    expect(getAnswerSpy).toHaveBeenCalledTimes(0);
  });

  test('escribir el simbolo "?" debe llamar a getAnswer', async () => {
    const input = wrapper.find(".indecision-container input");

    await input.setValue("Ask a question?");

    expect(getAnswerSpy).toHaveBeenCalledTimes(1);
  });

  test("el método getAnswer debe devolver un respuesta", async () => {
    await wrapper.vm.getAnswer();

    const img = wrapper.find("img");

    expect(img.exists()).toBeTruthy();

    //Answer return a random value contains in the hash/key "yes" of mapAnswer
    //We expect that in mapAnswer values exist the value wrapper.vm.answer

    expect(mapAnswer.yes).toContain(wrapper.vm.answer);
  });

  test("método getAnswer - fallo en llamada", async () => {
    fetch.mockImplementationOnce(() =>
      Promise.reject("API no está en servicio")
    );

    await wrapper.vm.getAnswer();

    const img = wrapper.find("img");

    expect(img.attributes("src")).toBe("/error.gif");

    expect(wrapper.vm.answer).toBe(
      "Ups... una pregunta muy existencial (Error API)"
    );
  });
});
