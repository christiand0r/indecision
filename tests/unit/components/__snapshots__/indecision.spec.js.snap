// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component debe ser igual al snapshot 1`] = `
<img src="/shiba_inu.gif" alt="Background">
<div class="indecision-container"><input type="text" name="question" id="question" placeholder="Escribe tu pregunta...">
  <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
  <div>
    <h2></h2>
    <!--v-if-->
  </div>
</div>
`;
